cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java

```diff
@@ -291,6 +310,19 @@ public void addOutput(String datasetName, Map<String, String> arguments) {
 
   @Override
   public void addOutput(String outputName, OutputFormatProvider outputFormatProvider) {
+    try {
```

Comment:
My feeling that this needs ti be little different: For every source or sink named, say, "xyz": You create a dataset "xyz.external". The properties of the dataset contain a list of sources and sinks that use that name. That way, multiple souurces/sinks can share the same name (and hence ext dataset) and you can still track them back to the actual sources/sinks that they represent. 

The current code would only record the first one with its configuration, but the second time, because the dataset already exists, the configuration will not be stored.